include/openPMD/Datatype.hpp

```diff
@@ -772,6 +771,182 @@ ReturnType switchType( Datatype dt, Action action, Args &&... args )
     }
 }
 
+namespace detail
+{
+// std::void_t is C++17
+template< typename >
+using void_t = void;
+
+/*
+ * Check whether class T has a member "errorMsg" of type std::string.
+ * Used to give helpful compile-time error messages with static_assert
+ * down in CallUndefinedDatatype.
+ */
+template< typename T, typename = void >
+struct HasErrorMessageMember
+{
+    static constexpr bool value = false;
+};
+
+template< typename T >
+struct HasErrorMessageMember<
+    T,
+    typename std::enable_if< std::is_same<
+        std::remove_cv_t< std::remove_reference_t< decltype( T::errorMsg ) > >,
+        std::string >::value >::type >
+{
+    static constexpr bool value = true;
+};
+
+/**
+ * Purpose of this struct is to detect at compile time whether
+ * Action::template operator()\<0\>() exists. If yes, call
+ * Action::template operator()\<n\>() with the passed arguments.
+ * If not, throw an error.
+ *
+ * @tparam n As in switchType().
+ * @tparam ReturnType As in switchType().
+ * @tparam Action As in switchType().
+ * @tparam Dummy For SFINAE, set to void.
```

Comment:
For [inclusive wording](https://developers.google.com/style/inclusive-documentation) we could replace _dummy_ with _placeholder_.